lib/vrf.c

```diff
@@ -384,54 +384,74 @@ static void vrf_hash_bitmap_free(void *data)
 	XFREE(MTYPE_VRF_BITMAP, bit);
 }
 
-vrf_bitmap_t vrf_bitmap_init(void)
+void vrf_bitmap_init(vrf_bitmap_t *pbmap)
 {
-	return hash_create_size(32, vrf_hash_bitmap_key, vrf_hash_bitmap_cmp,
-				"VRF BIT HASH");
+	*pbmap = NULL;
 }
 
-void vrf_bitmap_free(vrf_bitmap_t bmap)
+void vrf_bitmap_free(vrf_bitmap_t *pbmap)
 {
-	struct hash *vrf_hash = bmap;
+	struct hash *vrf_hash;
+
+	if (!*pbmap)
+		return;
+
+	vrf_hash = *pbmap;
 
 	hash_clean_and_free(&vrf_hash, vrf_hash_bitmap_free);
 }
 
-void vrf_bitmap_set(vrf_bitmap_t bmap, vrf_id_t vrf_id)
+void vrf_bitmap_set(vrf_bitmap_t *pbmap, vrf_id_t vrf_id)
 {
 	struct vrf_bit_set lookup = { .vrf_id = vrf_id };
-	struct hash *vrf_hash = bmap;
+	struct hash *vrf_hash;
 	struct vrf_bit_set *bit;
 
-	if (vrf_hash == NULL || vrf_id == VRF_UNKNOWN)
+	if (vrf_id == VRF_UNKNOWN)
 		return;
 
+	if (!*pbmap)
+		*pbmap = vrf_hash =
+			hash_create_size(2, vrf_hash_bitmap_key,
```

Comment:
a little concerned about the magic number '2', and about reducing the initial size so much. it'll take, what is it, four 'expand' cycles to get a hash back to the existing initial size?

Reply:
how often do you expect people to have > 16 vrf's configured on a box -vs- only having 1 or 2?  I would bet most of our use case is just having a couple, in which case 32 was way over kill

Reply:
hmm - well if that's true then ... why use a hash at all - maybe we should just use a simpler scheme with a smaller footprint?

Reply:
I'm not aware of anything else that would be simpler here.  o(1) lookups are pretty important imo.

I need to have a bit set for an arbitrary vrf id, which can range the entirety of the uint32_t space.  Original (naive) implementation created an actual bit based upon the ifindex value.  We ended up with bitmaps being absolutely huge( something like 64 mb per process per table ).  The original implementation over to a hash reduced it to several hundred k's of bytes.  I just want to reduce it further

Reply:
yup, ok, just wanted to ask